### ISSUE TITLE
seq2seq decoder ids

### ISSUE THREAD

**Issue by afuller187** [lucidrains/linear-attention-transformer#1]
Hi there, thanks for all the work. 

I am wondering how one should greedily sample the model in seq2seq. For example, lets say I want to do some MT. Is this how it works?

1. Get context (output of encoder) using src
2. Set tgt to be a start sequence token
3. Get logits using context and tgt
4. Find argmax of logits
5. Add the argmax of logits token to the end of tgt
6. Get new logits using context, and new tgt sequence, and repeat

So, tgt is essentially the decoder_ids (as in HuggingFace)

**Comment by lucidrains** [comment]
@afuller187 you are exactly right! I actually have a helper function on the AutoregressiveWrapper, which, once you wrap it around your decoder, helps you sample from the model, with either topk or nucleus sampling

https://github.com/lucidrains/linear-attention-transformer/blob/master/linear_attention_transformer/autoregressive_wrapper.py#L37

It is identical to https://github.com/lucidrains/reformer-pytorch#helpers-for-training-auto-regressive-models , just replace the `TrainingWrapper` with `AutoregressiveWrapper`

**Comment by lucidrains** [comment]
You can also look at this toy task as an example https://github.com/lucidrains/linear-attention-transformer/blob/master/examples/toy_tasks/copy_task.py#L81

**Comment by afuller187** [comment]
@lucidrains great thanks!

Sorry if this is a beginner question but I'm using these functions for your routing transformer:

enc_keys = encoder_output
yo = decoder(tgt, keys = enc_keys)

If I have 4 tokens in tgt, this will produce yo of shape (4, vocab_size). I'm assuming the last element of yo is the logits for my next token (which would be of shape 1 by vocab_size and I use for the argmax operation), and I can discard the first three. Is this correct?

Thanks again

**Comment by lucidrains** [comment]
@afuller187 that is correct! that's how decoding from an auto-regressive transformer currently works
